### ISSUE TITLE
how to load weights trained on multi gpu model to the single gpu model? 

### ISSUE THREAD

**Issue by Nandayang** [keras-team/keras#11253]
I trained on 4 gpus with multi_gpu model and saved weights with callbacks ModelCheckpoint, but when i want to run the model on single gpu machines, it seems i can't load the weights in to singel model. which get error: you are trying to load a weight file containing 1 layers into a model with 100 layers.(my single model is 100 layers).
After that ,i did as below:
```
parallel_model,single_model = XXnet()
parallel_model.load_weights(multi_gpu_weights.hdf5)
single_model .save("singe.h5")
```
It seems ok and there's no error when i load the single.h5,but when i use this to predict something ,errors happened: 
`terminate called after throwing an instance of 'std::bad_alloc'`
 it seems the python core has dumped.
Really need help, thanks!

**Comment by smthomas-sci** [comment]
One possibility...

```
def load_multigpu_checkpoint_weights(model, h5py_file):
    """
    Loads the weights of a weight checkpoint from a multi-gpu
    keras model.

    Input:

        model - keras model to load weights into

        h5py_file - path to the h5py weights file

    Output:
        None
    """

    print("Setting weights...")
    with h5py.File(h5py_file, "r") as file:

        # Get model subset in file - other layers are empty
        weight_file = file["model_1"]

        for layer in model.layers:

            try:
                layer_weights = weight_file[layer.name]

            except:
                # No weights saved for layer
                continue

            try:
                weights = []
                # Extract weights
                for term in layer_weights:
                    if isinstance(layer_weights[term], h5py.Dataset):
                        # Convert weights to numpy array and prepend to list
                        weights.insert(0, np.array(layer_weights[term]))

                # Load weights to model
                layer.set_weights(weights)

            except Exception as e:
                print("Error: Could not load weights for layer:", layer.name, file=stderr)
```

**Comment by ruian1** [comment]
Is this solved ?

**Comment by steerspace** [comment]
> One possibility...
> 
> ```
> def load_multigpu_checkpoint_weights(model, h5py_file):
>     """
>     Loads the weights of a weight checkpoint from a multi-gpu
>     keras model.
> 
>     Input:
> 
>         model - keras model to load weights into
> 
>         h5py_file - path to the h5py weights file
> 
>     Output:
>         None
>     """
> 
>     print("Setting weights...")
>     with h5py.File(h5py_file, "r") as file:
> 
>         # Get model subset in file - other layers are empty
>         weight_file = file["model_1"]
> 
>         for layer in model.layers:
> 
>             try:
>                 layer_weights = weight_file[layer.name]
> 
>             except:
>                 # No weights saved for layer
>                 continue
> 
>             try:
>                 weights = []
>                 # Extract weights
>                 for term in layer_weights:
>                     if isinstance(layer_weights[term], h5py.Dataset):
>                         # Convert weights to numpy array and prepend to list
>                         weights.insert(0, np.array(layer_weights[term]))
> 
>                 # Load weights to model
>                 layer.set_weights(weights)
> 
>             except Exception as e:
>                 print("Error: Could not load weights for layer:", layer.name, file=stderr)
> ```

Did you try this? I have a saved multi-gpu keras model, after used the function you provide, performance drops a lot. Dont know how to solve this. TAT

**Comment by smthomas-sci** [comment]
@steerspace Possibly it isn’t loading the batch normalisation weights if you had unlocked them to adjust them to your dataset. I stopped using this though as it’s a bit of a hack. The best work around is to create a custom model checkpoint callback that takes a reference to the original model, rather than the multi-gpu model as would an ordinary callback added to model.fit() / model.fit_generator. Check out my training code for my SemanticSegemenation repository, you’ll see an example. Works 100% of the time for me

**Comment by steerspace** [comment]
@smthomas-sci Custom model checkpoint callback works very well. 
The problem is that, if I only have a trained multi_gpu model weights file, how to extract single GPU model weights from it without retraining the network
I solved it by using this:
```python
model = ResNextImageNet(blah..)
p_model = multi_gpu_model(model)
p_model.load_weights(bst)  # load multi-gpu model weights
old_model = p_model.layers[-2]   #get single GPU model weights
# it's necessary to save the model before use this single GPU model
old_model.save(save_path)  
```
Thanks for your shares, I learned a lot.
btw, I am also working on histological images.:-)

**Comment by amapic** [comment]
@steermomo Thanks a lot

**Comment by solarflarefx** [comment]
> ```python
> p_model = multi_gpu_model(model)
> ```

So in the multi_gpu_model part, would the number of GPUs have to be specified?  I tried without specifying the number of gpus and got: 
For multi-gpu usage to be effective, call `multi_gpu_model` with `gpus >= 2`. Received: `gpus=1`

**Comment by proteus1991** [comment]
> > ```python
> > p_model = multi_gpu_model(model)
> > ```
> 
> So in the multi_gpu_model part, would the number of GPUs have to be specified? I tried without specifying the number of gpus and got: For multi-gpu usage to be effective, call `multi_gpu_model` with `gpus >= 2`. Received: `gpus=1`
> 
> The problem is, on the single gpu machine, if the number is set to > 1 then I get an error saying there are not enough GPUs on the system.
> 
> Does this mean that the saving of the single model would have to be done on the multi-GPU machine?

@solarflarefx have you solved this issue? Thanks!

**Comment by smthomas-sci** [comment]
@proteus1991 , this is related to loading weights on a keras model with a tensorflow v1 backend. As of tensorflow v2+, keras natively supports loading weights trained on multi-gpus, all by the simple `model.load_weights("path_to_model_weights..h5")` call.
